### PR TITLE
SQAUTHGITH-3 Add groups synchronisation

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java [hunk 1]
```diff
@@ -85,7 +89,7 @@ public boolean allowsUsersToSignUp() {
   public void init(InitContext context) {
     String state = context.generateCsrfState();
     OAuthService scribe = prepareScribe(context)
-      .scope("user:email")
+      .scope("user:email,read:org")
```
Comment (simonbrandhof) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: The scope `read:org` must be requested only if synchronization of groups is enabled.

Comment (julienlancelot) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: I was asking myself this question, but if the option is enabled then disabled then enabled, the user will have to validate each time new permissions.
But that's probably a corner case as it should never happen...

Comment (simonbrandhof) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: This would be the expected behavior for end-users, so it makes sense.

Comment (julienlancelot) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: ok

**Comment by julienlancelot** [comment]
Feedbacks applied

File: src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java [hunk 2]
```diff
@@ -101,28 +105,47 @@ public void callback(CallbackContext context) {
...
+    return GsonUser.parse(responseBody);
+  }
+
+  private List<GsonTeams.GsonTeam> getTeams(OAuthService scribe, Token accessToken) {
+    String responseBody = executeRequest("https://api.github.com/user/teams", scribe, accessToken);
+    LOGGER.trace("Teams response received : {}", responseBody);
+    return GsonTeams.parse(responseBody);
+  }
+
+  private String executeRequest(String requestUrl, OAuthService scribe, Token accessToken) {
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: MAJOR Make "executeRequest" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java [hunk 3]
```diff
@@ -153,4 +176,14 @@ private static String getName(GsonUser gsonUser) {
   private String generateUniqueLogin(GsonUser gsonUser) {
     return getKey() + "@" + gsonUser.getLogin();
   }
+
+  private enum TeamToGroup implements Function<GsonTeams.GsonTeam, String> {
+    INSTANCE;
+
+    @Override
+    public String apply(@Nonnull GsonTeams.GsonTeam gsonTeam) {
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java:184]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: src/main/java/org/sonarsource/auth/github/GsonTeams.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Lite representation of JSON response of GET https://api.github.com/user/orgs
+ */
+public class GsonTeams {
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/auth/github/GsonTeams.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1118)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java [hunk 5]
```diff
@@ -101,28 +105,47 @@ public void callback(CallbackContext context) {
...
     context.redirectToRequestedPage();
   }
 
+  private GsonUser getUser(OAuthService scribe, Token accessToken) {
+    String responseBody = executeRequest("https://api.github.com/user", scribe, accessToken);
+    LOGGER.trace("User response received : {}", responseBody);
+    return GsonUser.parse(responseBody);
+  }
+
+  private List<GsonTeams.GsonTeam> getTeams(OAuthService scribe, Token accessToken) {
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: MAJOR Make "getTeams" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java [hunk 6]
```diff
@@ -101,28 +105,47 @@ public void callback(CallbackContext context) {
...
+    if (settings.syncGroups()) {
+      List<GsonTeams.GsonTeam> teams = getTeams(scribe, accessToken);
+      builder.setGroups(from(teams).transform(TeamToGroup.INSTANCE).toSet());
+    }
+
+    context.authenticate(builder.build());
     context.redirectToRequestedPage();
   }
 
+  private GsonUser getUser(OAuthService scribe, Token accessToken) {
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/auth/github/GitHubIdentityProvider.java]: MAJOR Make "getUser" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by simonbrandhof** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
- MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by julienlancelot** [comment]
Must wait for release of version 1.1 before merging to master

**Pull request merged**